### PR TITLE
SONARJAVA-1506 Handle widlcards from sources

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-checks/src/test/files/checks/CollectionInappropriateCallsCheck.java [hunk 1]
```diff
@@ -140,3 +140,12 @@ void f() {
   }
 
 }
+
+class G {
+  public <E> void union(final Set<? extends E> set1, Object o) {
+    boolean trux = set1.contains(o); // Noncompliant
```
Comment (benzonico) [inline java-checks/src/test/files/checks/CollectionInappropriateCallsCheck.java]: huh? o can be a E and so this code would be totally compliant wouldn't it ?

File: java-squid/src/main/java/org/sonar/java/resolve/JavaSymbol.java [hunk 2]
```diff
@@ -679,4 +679,38 @@ public LabeledStatementTree declaration() {
 
   }
 
+  public static class WidlcardSymbol extends TypeJavaSymbol {
```
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/resolve/JavaSymbol.java]: typo!

File: java-squid/src/main/java/org/sonar/java/resolve/JavaType.java [hunk 3]
```diff
@@ -384,5 +397,122 @@ public JavaType substitution(TypeVariableJavaType typeVariableType) {
...
+    private final boolean unbunded;
+
+    public WildCardType(String name, JavaType bound, boolean unbounded) {
+      super(new JavaSymbol.WidlcardSymbol(name));
+      this.bound = bound;
+      this.unbunded = unbounded;
+      this.symbol.type = this;
+    }
+
+    public boolean isUnbunded() {
```
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/resolve/JavaType.java]: typo?

File: java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java [hunk 4]
```diff
@@ -101,9 +104,49 @@ private static boolean isRedundantCast(JavaType cast, JavaType expressionType) {
     if(erasedExpressionType.isTagged(JavaType.TYPEVAR)) {
       erasedExpressionType = erasedExpressionType.erasure();
     }
+    // FIXME SOANRJAVA-1514: all this part should be replaced by :
+    // return erasedExpressionType.equals(cast) || (!cast.isNumerical() && erasedExpressionType.isSubtypeOf(cast));
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java]: MAJOR This block of commented-out lines of code should be removed. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java [hunk 5]
```diff
@@ -24,9 +24,11 @@
 import org.sonar.api.server.rule.RulesDefinition;
 import org.sonar.check.Priority;
 import org.sonar.check.Rule;
+import org.sonar.java.resolve.JavaSymbol.WidlcardSymbol;
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java]: MINOR Remove this unused import 'org.sonar.java.resolve.JavaSymbol.WidlcardSymbol'. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUselessImportCheck)

File: java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java [hunk 6]
```diff
@@ -101,9 +104,49 @@ private static boolean isRedundantCast(JavaType cast, JavaType expressionType) {
     if(erasedExpressionType.isTagged(JavaType.TYPEVAR)) {
       erasedExpressionType = erasedExpressionType.erasure();
     }
+    // FIXME SOANRJAVA-1514: all this part should be replaced by :
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1134)

File: java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java [hunk 7]
```diff
@@ -24,9 +24,11 @@
 import org.sonar.api.server.rule.RulesDefinition;
 import org.sonar.check.Priority;
 import org.sonar.check.Rule;
+import org.sonar.java.resolve.JavaSymbol.WidlcardSymbol;
 import org.sonar.java.resolve.JavaType;
 import org.sonar.java.tag.Tag;
 import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.semantic.Symbol.TypeSymbol;
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java]: MINOR Remove this unused import 'org.sonar.plugins.java.api.semantic.Symbol.TypeSymbol'. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUselessImportCheck)

File: java-squid/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java [hunk 8]
```diff
@@ -358,12 +358,21 @@ public void visitParameterizedType(ParameterizedTypeTree tree) {
...
+    if (!unbounded) {
       resolveAs(tree.bound(), JavaSymbol.TYP);
-      registerType(tree, getType(tree.bound()));
+      bound = getType(tree.bound());
+    }
+    registerType(tree, newWildcardType(tree, bound, unbounded));
+  }
+
+  private JavaType.WildCardType newWildcardType(WildcardTree tree, JavaType bound, boolean unbunded) {
+    String name = unbunded ? "?" : "? " + tree.extendsOrSuperToken().text() + " " + tree.bound().symbolType().name();
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java]: MAJOR Add parentheses to make the operator precedence explicit. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS864)

File: java-squid/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java [hunk 9]
```diff
@@ -358,12 +358,21 @@ public void visitParameterizedType(ParameterizedTypeTree tree) {
...
+    boolean unbounded = tree.is(Tree.Kind.UNBOUNDED_WILDCARD);
+    if (!unbounded) {
       resolveAs(tree.bound(), JavaSymbol.TYP);
-      registerType(tree, getType(tree.bound()));
+      bound = getType(tree.bound());
+    }
+    registerType(tree, newWildcardType(tree, bound, unbounded));
+  }
+
+  private JavaType.WildCardType newWildcardType(WildcardTree tree, JavaType bound, boolean unbunded) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java]: MAJOR Make "newWildcardType" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 7 issues:
* MAJOR 5 major
* MINOR 2 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "isSubtypeOf" is 13 which is greater than 10 authorized.](https://github.com/SonarSource/sonar-java/blob/79a44527f8c9004378922d6c00c61c33ec9b93ba/java-squid/src/main/java/org/sonar/java/resolve/JavaType.java#L240) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java (hunk 6 above)
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java]: please precise why...

File: java-checks/src/main/java/org/sonar/java/checks/CollectionInappropriateCallsCheck.java [hunk 10]
```diff
@@ -100,6 +100,9 @@ private static Type getTypeParameter(Type collectionType) {
   }
 
   private static boolean isArgumentCompatible(Type argumentType, Type collectionParameterType) {
+    if (collectionParameterType instanceof JavaType.WildCardType) {
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/CollectionInappropriateCallsCheck.java]: This lacks a fixme with a link to the ticket and an explanation.

File: java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java [hunk 11]
```diff
@@ -24,9 +24,11 @@
 import org.sonar.api.server.rule.RulesDefinition;
 import org.sonar.check.Priority;
 import org.sonar.check.Rule;
+import org.sonar.java.resolve.JavaSymbol.WildcardSymbol;
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java]: MINOR Remove this unused import 'org.sonar.java.resolve.JavaSymbol.WildcardSymbol'. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUselessImportCheck)

File: java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java [hunk 12]
```diff
@@ -24,9 +24,11 @@
...
 import org.sonar.java.resolve.JavaType;
 import org.sonar.java.tag.Tag;
 import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.semantic.Symbol.TypeSymbol;
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java]: MINOR Remove this unused import 'org.sonar.plugins.java.api.semantic.Symbol.TypeSymbol'. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 3 major
* MINOR 2 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "isSubtypeOf" is 13 which is greater than 10 authorized.](https://github.com/SonarSource/sonar-java/blob/6648215c22f757f6d963111b3a43b7fff94186f5/java-squid/src/main/java/org/sonar/java/resolve/JavaType.java#L241) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 3 major
* MINOR 2 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "isSubtypeOf" is 13 which is greater than 10 authorized.](https://github.com/SonarSource/sonar-java/blob/1d286f48a552d762c8f775be3dbda32dcd28190b/java-squid/src/main/java/org/sonar/java/resolve/JavaType.java#L241) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: java-squid/src/main/java/org/sonar/java/resolve/JavaType.java [hunk 13]
```diff
@@ -384,5 +396,111 @@ public JavaType substitution(TypeVariableJavaType typeVariableType) {
...
+      return boundType == BoundType.UNBOUNDED;
+    }
+
+    @Override
+    public boolean isSubtypeOf(String fullyQualifiedName) {
+      return false;
+    }
+
+    @Override
+    public boolean isSubtypeOf(Type superType) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/JavaType.java]: MAJOR The Cyclomatic Complexity of this method "isSubtypeOf" is 11 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "isSubtypeOf" is 13 which is greater than 10 authorized.](https://github.com/SonarSource/sonar-java/blob/d1da8488babc7ea347f4a9b0aa42f90e700e6e62/java-squid/src/main/java/org/sonar/java/resolve/JavaType.java#L239) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: java-squid/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 14]
```diff
@@ -509,19 +515,23 @@ private boolean isArgumentsAcceptable(List<JavaType> argTypes, List<JavaType> fo
...
+      if (!isAcceptableType(argTypes.get(i), formals.get(i), autoboxing, isParameterized)) {
         return false;
       }
     }
     return true;
   }
 
-  private boolean isAcceptableType(JavaType arg, JavaType formal, boolean autoboxing) {
+  private boolean isAcceptableType(JavaType arg, JavaType formal, boolean autoboxing, boolean isParameterized) {
+    // FIXME Inference of types for method using parameter types is not handled so fall back to behavior ignoring wildcards
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/Resolve.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1134)

File: java-checks/src/main/java/org/sonar/java/checks/CollectionInappropriateCallsCheck.java [hunk 15]
```diff
@@ -100,6 +100,10 @@ private static Type getTypeParameter(Type collectionType) {
   }
 
   private static boolean isArgumentCompatible(Type argumentType, Type collectionParameterType) {
+    // FIXME SONARJAVA-1514 subtyping of type using generics is not handled correctly, but wildcards are covered
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/CollectionInappropriateCallsCheck.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1134)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* MAJOR 4 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "isSubtypeOf" is 13 which is greater than 10 authorized.](https://github.com/SonarSource/sonar-java/blob/96a4717a2ff752512afe3e76d6e5db92ece2852a/java-squid/src/main/java/org/sonar/java/resolve/JavaType.java#L239) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* MAJOR 4 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "isSubtypeOf" is 13 which is greater than 10 authorized.](https://github.com/SonarSource/sonar-java/blob/96a4717a2ff752512afe3e76d6e5db92ece2852a/java-squid/src/main/java/org/sonar/java/resolve/JavaType.java#L239) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* MAJOR 4 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "isSubtypeOf" is 13 which is greater than 10 authorized.](https://github.com/SonarSource/sonar-java/blob/c15fd799bdacbd8fa1fa09d86a9d2f6807a2b285/java-squid/src/main/java/org/sonar/java/resolve/JavaType.java#L239) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: java-squid/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 16]
```diff
@@ -509,22 +516,30 @@ private boolean isArgumentsAcceptable(List<JavaType> argTypes, List<JavaType> fo
...
+      if (!isAcceptableType(argTypes.get(i), formals.get(i), autoboxing, isParameterized)) {
         return false;
       }
     }
     return true;
   }
 
-  private boolean isAcceptableType(JavaType arg, JavaType formal, boolean autoboxing) {
+  private boolean isAcceptableType(JavaType arg, JavaType formal, boolean autoboxing, boolean isParameterized) {
+    // FIXME Inference of types for method using parameter types is not handled so fall back to behavior based on erasure
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/Resolve.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1134)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* MAJOR 4 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "isSubtypeOf" is 13 which is greater than 10 authorized.](https://github.com/SonarSource/sonar-java/blob/d8788fba004f277cf0a01ae738ac9d1d92e42aaa/java-squid/src/main/java/org/sonar/java/resolve/JavaType.java#L235) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* MAJOR 4 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "isSubtypeOf" is 13 which is greater than 10 authorized.](https://github.com/SonarSource/sonar-java/blob/decdbc052966b5af8db4e1c4d3dd214a21d6a25f/java-squid/src/main/java/org/sonar/java/resolve/JavaType.java#L235) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Pull request merged**